tutorials/shaders/shader_reference/shader_preprocessor.rst

```diff
@@ -1,56 +1,256 @@
+
 #if
 ^^^
-\ **Syntax:** `#if condition`.
 
-The `#if` directive checks the condition and if it evaluates to a non-zero value, the code block is included, otherwise it is skipped.
-In order to evaluate, the condition must be an expression giving a simple floating-point, integer or boolean result. There may be multiple condition blocks connected by `||` or `&&` operators.
-It may be continued by a `#else` block, but must be ended with the `#endif` directive.
+**Syntax:** ``#if <condition>``
+
+The ``#if`` directive checks the ``condition`` passed. If it evaluates to a non-zero
+value, the code block is included, otherwise it is skipped.
+
+To evaluate correctly, the condition must be an expression giving a simple
+floating-point, integer or boolean result. There may be multiple condition
+blocks connected by ``&&`` (AND) or ``||`` (OR) operators. It may be continued
+by a ``#else`` block, but **must** be ended with the ``#endif`` directive.
+
+.. code-block:: glsl
+
+    #define VAR 3
+    #define USE_LIGHT 0 // Evaluates to `false`.
+    #define USE_COLOR 1 // Evaluates to `true`.
+
+    #if VAR == 3 && (USE_LIGHT || USE_COLOR)
+    // Condition is `true`. Include this portion in the final shader.
+    #endif
+
+Using the ``defined()`` *preprocessor function*, you can check whether the
+passed identifier is defined a by ``#define`` placed above that directive. This
+is useful for creating multiple shader versions in the same file. It may be
+continued by a `#else` block, but must be ended with the ``#endif`` directive.
+
+The ``defined()`` function's result can be negated by using the ``!`` (boolean NOT)
+symbol in front of it. This can be used to check whether a define is *not* set.
+
+.. code-block:: glsl
+
+    #define USE_LIGHT
+    #define USE_COLOR
+
+    // Correct syntax:
+    #if defined(USE_LIGHT) || defined(USE_COLOR) || !defined(USE_REFRACTION)
+    // Condition is `true`. Include this portion in the final shader.
+    #endif
+
+Be careful, as ``defined()`` must only wrap a single identifier within parentheses, never more:
+
+.. code-block:: glsl
+
+    // Incorrect syntax (parentheses are not placed where they should be):
+    #if defined(USE_LIGHT || USE_COLOR || !USE_REFRACTION)
+    // This will cause an error or not behave as expected.
+    #endif
+
+.. tip::
+
+    In the shader editor, preprocessor branches that evaluate to ``false`` (and
+    are therefore excluded from the final compiled shader) will appear grayed
+    out. This does not apply to run-time ``if`` statements.
+
+**#if preprocessor versus if statement: Performance caveats**
+
+The :ref:`shading language <doc_shading_language>` supports run-time ``if`` statements:
+
+.. code-block:: glsl
+
+    const bool USE_LIGHT = true;
+
+    if (USE_LIGHT) {
+        // This part is included in the compiled shader, and always run.
+    } else {
+        // This part is included in the compiled shader, but never run.
```

Comment:
This isn't true, since USE_LIGHT is a compile time constant this code will get compiled out. Perhaps the example would be better using a uniform?

Also note, your discussion of performance below is only really applicable if USE_LIGHT is a uniform or a run time variable

Reply:
I've changed it to use a uniform.